### PR TITLE
ensure uint8 is honoured for cpu operations in dynamo

### LINKED ISSUE #105264: Inductor generates incorrect CPU code for `uint8` operations
### 🐛 Describe the bug

Consider the following snippet, which adds a `torch.uint8` tensor to itself, _then_ casts the result to `torch.int16`:

```python
import torch

def f(x):
    return (x + x).to(torch.int16)

x = torch.tensor(128, dtype=torch.uint8)
print(f(x))
print(torch.compile(f)(x))
```

This produces the following output (`0` is correct):

```python
tensor(0, dtype=torch.int16)
tensor(256, dtype=torch.int16)
```

The relevant generated CPU kernel:

```cpp
extern "C" void kernel(const unsigned char* in_ptr0,
                       short* out_ptr0)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        auto tmp1 = tmp0 + tmp0;
        auto tmp2 = static_cast<short>(tmp1);
        out_ptr0[static_cast<long>(0L)] = tmp2;
    }
}
```

### Error logs

_No response_

### Minified repro

_No response_

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230714+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.9.5 (default, Nov 23 2021, 15:27:38)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-148-generic-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 530.30.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 57 bits virtual
CPU(s):                          32
On-line CPU(s) list:             0-31
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           106
Model name:                      Intel(R) Xeon(R) Silver 4314 CPU @ 2.40GHz
Stepping:                        6
Frequency boost:                 enabled
CPU MHz:                         3136.478
CPU max MHz:                     3400.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4800.00
Virtualization:                  VT-x
L1d cache:                       768 KiB
L1i cache:                       512 KiB
L2 cache:                        20 MiB
L3 cache:                        24 MiB
NUMA node0 CPU(s):               0-31
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.25.1
[pip3] torch==2.1.0.dev20230714+cpu
[conda] No relevant packages
```

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

### PR DESCRIPTION
**Pull request by davidradl**
Fixes #105264 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Review by lezcano — COMMENTED** [review]
Please put the tests in `test_torchinductor.py`. See the tests there for examples on how to write them.

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -392,6 +392,9 @@ def inner_fn(index):
                 return override_fn_when_input_bool(*[load(index) for load in loaders])
             elif override_fn_when_cuda_float64 and is_cuda and dtype == torch.float64:
                 return override_fn_when_cuda_float64(*[load(index) for load in loaders])
+            elif dtype == torch.uint8 or dtype == torch.int8 or dtype == torch.int16 or dtype == torch.int32:
```
Comment (lezcano) [inline torch/_inductor/lowering.py:395]: This should be only if `not is_cuda`. A nice follow-up would be to put together tests, and if necessary, a similar fix for CUDA, if triton has C++ semantics rather than Python semantics (not sure about this).

[the remark this answers is not in the document]
Comment (davidradl) [reply torch/_inductor/lowering.py:395]: > Why `int32`? The problem with smaller types is that the C++ "usual arithmetic conversions" promote them to `int`, but `int32` shouldn't have any extraneous promotion.

Thanks for your feedback. There is a problem that the test case shows - it sounds like the fix should not be here - I will remove the int32 test and raise another issue to track that separately.

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -7712,6 +7712,48 @@ def fn(pytype, dtype):
...
+                return (x * x).to(torch.int16)
+            def mySquare16(x):
+                return (x ** 2).to(torch.int16)
+
+            x = torch.tensor(128, dtype=torch.uint8)
+
+            torchResult = mySum16(x)
+            dynamoResult = torch.compile(mySum16)(x)
+
+            assert(torchResult == dynamoResult == 0)
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:7732]: Rather than doing this, see `self.common`, which also provides a few other tests behind the hood.

File: test/inductor/test_torchinductor.py [hunk 3]
```diff
@@ -7712,6 +7712,48 @@ def fn(pytype, dtype):
...
+
+            x = torch.tensor(120, dtype=torch.int8)
+            torchResult = mySum16(x)
+            dynamoResult = torch.compile(mySum16)(x)
+
+            assert(torchResult == dynamoResult == -16)
+            def mySum32(x):
+                return (x + x).to(torch.int32)
+
+            x = torch.tensor( 35000, dtype=torch.int32)
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:7752]: What's this example testing?

Comment (davidradl) [reply test/inductor/test_torchinductor.py:7752]: good spot - it is not needed - I will remove

**Review by jansel — CHANGES_REQUESTED** [review]

File: test/inductor/test_torchinductor.py (hunk 2 above)
Comment (davidradl) [reply test/inductor/test_torchinductor.py:7732]: > Rather than doing this, see `self.common`, which also provides a few other tests behind the hood.

@lezcano I have had a look at this. I see there is a common template that is used to run CPU and CUDA tests.
I have a mac system, this `if` does not run as  torch.backends.mps.is_available() is true. Commenting out the mps check gives me errors.

```
 if HAS_CPU and not torch.backends.mps.is_available():

    class SweepInputsCpuTest(SweepInputs2, TestCase):
        gen = InputGen(10, "cpu")

    SweepInputsCpuTest.populate()

    class CpuTests(TestCase):
        common = check_model
        device = "cpu"


    copy_tests(CommonTemplate, CpuTests, "cpu") 

```
I will make the changes and see if the pr build works with the tests in the new place

File: test/inductor/test_torchinductor.py [hunk 4]
```diff
@@ -7712,6 +7712,48 @@ def fn(pytype, dtype):
 
                 self.assertEqual(ret_opt, fn(pytype, dtype))
 
+        # Test that the compiler and torch output for binary operations match.
+        # Particularly check scenarios where the operation performed would result in 'wrapping',
+        # i.e., when the result would be too big for the type so a negative value is returned.
+        def test_lowering(self):
+
+            def mySum16(x):
```
[the remark this answers is not in the document]
Comment (davidradl) [reply test/inductor/test_torchinductor.py]: I have moved the tests into standard template driven part, as per the other review comment, if we decide we need to keep the tests here - I will rename the methods names as per your suggestion.

File: test/inductor/test_torchinductor.py [hunk 5]
```diff
@@ -604,6 +604,56 @@ def fn(x, y, z):
...
+
+    def test_square_uint8_tensor(self):
+        def fn(x):
+            return (x ** 2).to(torch.int16)
+        x = torch.tensor(8, dtype=torch.uint8)
+
+        self.common(fn, x)
+
+    #  issue https://github.com/pytorch/pytorch/issues/108520 raised for this
+    # def test_uint32_tensor(self):
```
Comment (jansel) [inline test/inductor/test_torchinductor.py:651]:
```suggestion
    # def test_int32_tensor(self):
```

**Comment by jansel** [comment]
Looks like tests are failing.  You might need to use `torch.tensor([128])` so they are 1D.

`make setup_lint && lintrunner -a` to fix lints.

**Comment by davidradl** [comment]
> Looks like tests are failing. You might need to use `torch.tensor([128])` so they are 1D.
> 
> `make setup_lint && lintrunner -a` to fix lints.

thanks I will try that. If the CPU ones work and the CUDA ones do not, then I will revert to the local tests outside of the templated tests, as I know they work for me locally (on a mac anyway!)

**Comment by jansel** [comment]
If cuda fails (not sure if it will), then the same fix will likely work on cuda too.

File: test/inductor/test_torchinductor.py (hunk 2 above)
Comment (davidradl) [reply test/inductor/test_torchinductor.py]: the tests have all passed in the template place :-)

File: test/inductor/test_torchinductor.py [hunk 6]
```diff
@@ -604,6 +604,54 @@ def fn(x, y, z):
...
+
+        x = torch.tensor([35000], dtype=torch.int16)
+
+        self.common(fn, x)
+
+    def test_multiply_uint8_tensor(self):
+        def fn(x):
+            return (x + x).to(torch.int16)
+
+        x = torch.tensor([8], dtype=torch.uint8)
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:636]:
```suggestion
        x = torch.tensor([100], dtype=torch.uint8)
```

File: test/inductor/test_torchinductor.py [hunk 7]
```diff
@@ -604,6 +604,54 @@ def fn(x, y, z):
...
+
+        self.common(fn, x)
+
+    def test_square_uint8_tensor(self):
+        def fn(x):
+            return (x**2).to(torch.int16)
+
+        x = torch.tensor([8], dtype=torch.uint8)
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:644]:
```suggestion
        x = torch.tensor([32], dtype=torch.uint8)
```

File: test/inductor/test_torchinductor.py (hunk 6 above)
Comment (davidradl) [reply test/inductor/test_torchinductor.py:636]: What is the thinking in changing the number here. I was trying to get it to 256 which is 0. I could add anotehr test with 100 to check a non zero the over flow case. Or have I misunderstood?

File: test/inductor/test_torchinductor.py (hunk 7 above)
Comment (peterbell10) [reply test/inductor/test_torchinductor.py:644]: The square root of 256 is 16, not 8. Also (32 * 32 % 256) is also 0.

Comment (lezcano) [reply test/inductor/test_torchinductor.py:644]: ops, I wanted to put any large number, but "32" is not a great one. My point was that 8**2 does not overflow here, so it's always going to be correct unless we do something very wrong

Comment (davidradl) [reply test/inductor/test_torchinductor.py:644]: @peterbell10 @lezcano oops.
 I will correct!

**Comment by lezcano** [comment]
Tests are still failing.

**Comment by davidradl** [comment]
It looks like the cpu tests fail: 
```
test_add_int16_tensor_cpu and gpu fail with: value cannot be converted to type int16 without overflow
test_add_int8_tensor_dynamic_shapes_cpu - AssertionError: False is not true : Failed to find 
test_add_uint8_tensor_dynamic_shapes_cpu - AssertionError: False is not true : Failed to find 
```

but  
```
test_add_int8_tensor_cpu
test_add_uint8_tensor_cpu
```
appear to work (just the dynamic ones fail)

Investigating further locally - it might be we want to amend the logic to only apply to the ones we know fail - Clang on M1 is failing for me.

File: test/inductor/test_torchinductor.py [hunk 8]
```diff
@@ -604,6 +604,70 @@ def fn(x, y, z):
...
+
+        x = torch.tensor([120], dtype=torch.int8)
+
+        self.common(fn, x)
+
+    def test_add_int16_tensor(self):
+        def fn(x):
+            return (x + x).to(torch.int32)
+
+        x = torch.tensor([35000], dtype=torch.int16)
```
Comment (peterbell10) [inline test/inductor/test_torchinductor.py:628]: The error is right, `32767` is the maximum possible `int16` value. This error is coming from eager.

**Comment by lezcano** [comment]
Could you make sure all the tests added in this PR and failing in a previous CI run pass locally before pushing to the PR? That would save plenty of CI cycles, which are not particularly cheap.

File: test/inductor/test_torchinductor.py (hunk 8 above)
Comment (davidradl) [reply test/inductor/test_torchinductor.py]: Thanks - another mistake corrected.

**Comment by davidradl** [comment]
> Could you make sure all the tests added in this PR and failing in a previous CI run pass locally before pushing to the PR? That would save plenty of CI cycles, which are not particularly cheap.

I agree it is frustrating testing blind. I have a Mac and I performed testing that showed the fix working for me. The test code I added, was not actually driven for a mac only Linux.  I am not sure what the easiest way is to run these Linux, CPU and CUDA tests prior to the PR build, what would you suggest?

**Comment by lezcano** [comment]
The support for mac in PyTorch is a tad spotty, when it comes to general development. It'd be best to get your hands on a linux box, either physically or through aws/azure/gcp or a variation of these.

**Comment by davidradl** [comment]
> The support for mac in PyTorch is a tad spotty, when it comes to general development. It'd be best to get your hands on a linux box, either physically or through aws/azure/gcp or a variation of these.

thanks I will see what I can do.

**Comment by peterbell10** [comment]
@davidradl that test is automatically generated from the test cases you've added here. As you noticed, it inspects loop variables in the generated code, but your example only uses single element tensors so it can't find any. This is benign and you can simply skip the test by adding it to the `test_failures` list in `test_torchinductor_codegen_dynamic_shapes.py`.

File: test/inductor/test_torchinductor.py [hunk 9]
```diff
@@ -604,6 +604,70 @@ def fn(x, y, z):
 
         self.common(fn, (x, y, z))
 
+    # lowering operation tests
+    def test_add_uint8_tensor(self):
+        def fn(x):
+            return (x + x).to(torch.int16)
+
+        x = torch.tensor([128], dtype=torch.uint8)
```
Comment (jansel) [inline test/inductor/test_torchinductor.py:612]: Maybe you need to do something like:
```suggestion
        x = torch.tensor([128]*10, dtype=torch.uint8)
```
to fix the tests checking for a loop.

For only 1 element I think inductor might optimize away the loop.


Either that or update the test to check something better.

**Review by jansel — CHANGES_REQUESTED** [review]

File: test/inductor/test_torchinductor.py (hunk 9 above)
Comment (davidradl) [reply test/inductor/test_torchinductor.py:612]: It seems I can either:

- remove the test using the test_failures
or
- change the test to avoid the error by intruding a second element.

Both of these seem like circumventions, whcih we should do if we have to.  
Could we make the dynamic test more resilient so that if it gets no loop variable then it does not generate test code that would fail.  Maybe something like:

```
lines = code.split("\n")
if lines && lines.len() >0:
        for line in lines:
            if "for(" in line:
                for_loop_found = True
                if re.search(r";.*ks.*;", line) is not None:
                    has_dynamic = True
                    break
        self.assertTrue(
            has_dynamic, msg=f"Failed to find dynamic for loop variable\n{code}"
```

Or do we think this might mask bugs where no output occurs.

Comment (davidradl) [reply test/inductor/test_torchinductor.py:612]: I will fix the way that @jansel recommends - and raise an issue for the single parameter version

Comment (davidradl) [reply test/inductor/test_torchinductor.py]: Unfortunately  `x = torch.tensor([9]*2, dtype=torch.uint8)` files with errors like 
`FAILED [0.0009s] inductor/test_torchinductor.py::CpuTests::test_add_int16_tensor_cpu - TypeError: CommonTemplate.test_add_int16_tensor.<locals>.fn() takes 1 positional argument but 2 were given` . I am trying bracketing the operation.

File: test/inductor/test_torchinductor.py [hunk 10]
```diff
@@ -604,6 +604,99 @@ def fn(x, y, z):
...
+        def fn(x):
+            return (x + x).to(torch.int16)
+
+        # issue 109016 has been raised to track that the below line fails the dynamic shape test
+        # x = torch.tensor([128], dtype=torch.uint8)
+        # issue 109020 has been raised to track that the below line fails with 'takes 1 positional argument but 2 were given'
+        # x = torch.tensor(([64]*20), dtype=torch.uint8)
+        x = torch.tensor([128], dtype=torch.uint8)
+
+        self.common(fn, x)
```
Comment (peterbell10) [inline test/inductor/test_torchinductor.py:618]: This function takes a list, so when you call it with a tensor it coerces the tensor into a list of tensors. When the length of the leading dimension is larger than 1, it's interpreted as multiple tensor arguments. Hence, the error "fn takes 1 positional argument but 2 were given". The fix is to pass the arguments as a list.
```suggestion
        self.common(fn, (x,))
```

Comment (davidradl) [reply test/inductor/test_torchinductor.py:618]: thanks @peterbell10 I had not noticed that.

Comment (davidradl) [reply test/inductor/test_torchinductor.py]: @peterbell10 thanks for all your help. I now see  `2 New Failures, 2 Unrelated Failures` what do you suggest? They do not seem related to my change.

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
Might want to try rebasing to ensure the failures are related.

**Comment by davidradl** [comment]
> Might want to try rebasing to ensure the failures are related.
@jansel I ran:
```
git rebase upstream/main
Current branch git105264 is up to date.
(base) davidradley@Davids-MBP-2 pytorch % git status
On branch git105264
Your branch is up to date with 'origin/git105264'.

nothing to commit, working tree clean
(base) davidradley@Davids-MBP-2 pytorch % 
```
It looks up to date.

**Comment by lezcano** [comment]
you are missing a `git fetch upstream`

**Comment by davidradl** [comment]
> git fetch upstream

Thanks .

**Comment by davidradl** [comment]
The rebase pulled in a lot of commits. I am not sure how to get rid of them, so I will raise another pr with just my changes.
